bldsys/cmake/module/FindVulkan.cmake

```diff
@@ -233,6 +233,17 @@ if("FATAL_ERROR" IN_LIST Vulkan_FIND_COMPONENTS)
     list(REMOVE_ITEM Vulkan_FIND_COMPONENTS "FATAL_ERROR")
 endif()
 
+# FindVulkan only works correctly with the default CMAKE_FIND_FRAMEWORK
+# value: FIRST. If LAST it will find the macOS dylibs instead of the iOS
+# frameworks when IOS is true. If ALWAYS it will fail to find the macOS
+# dylibs. If NEVER it will fail to find the iOS frameworks. If frameworks
+# are ever included in the SDK for macOS, the search mechanism will need
+# revisiting.
```

Comment:
Not sure this is correct.  When I run VULKAN_SDK/ios/setup_env.sh the find_vulkan.cmake resolves correctly as currently is deployed.  The CI environment also builds/resolves correctly and locally, it resolves correctly.  Could you confirm that your changes are necessary?

Reply:
In the case of iOS, FindVulkan sets the library search path it uses to `$VULKAN_SDK/lib; $VULKAN_SDK/../iOS/lib`. If you use VULKAN_SDK/ios/setup_env.sh to set VULKAN_SDK then it ends up with the search path `VULKAN_SDK/iOS/lib; VULKAN_SDK/iOS/lib`. In that case it will not be affected by the setting of CMAKE_FIND_FRAMEWORK, _unless_ you have installed the macOS libraries globally.<sup>1</sup>

However if you use VULKAN_SDK/setup-env.sh to set your environment then `$VULKAN_SDK` will be VULKAN_SDK/macOS and FindVulkan will use the search path `VULKAN_SDK/macOS/lib; VULKAN_SDK/iOS/lib`. CMAKE_FIND_FRAMEWORK is always in play in this case.

For anyone building for both macOS and iOS having to have different values of VULKAN_SDK does not work. It needs to be fixed. My choice is to use VULKAN_SDK/setup-env.sh as everything except the libraries and `share/vulkan` is common to both. This results in the searcy path in the previous paragraph. Setting CMAKE_FIND_FRAMEWORK ensures the correct library is found. It does not break the case when VULKAN_SDK/ios/setup_env.sh was used. In fact it makes it more robust by guaranteeing to find frameworks first<sup>1</sup>. For `share/vulkan` my cmake code uses `${Vulkan_Target_SDK}/${CMAKE_SYSTEM_NAME}/share/vulkan` to ensure it gets the correct files. `${Vulkan_Target_SDK}` is set by FindVulkan to the root of the SDK.

<sup>1</sup> CMake goes through its entire search sequence looking for a dylib or framework as indicated by CMAKE_FIND_FRAMEWORK. If it doesn't find anything it repeats the search from the beginning of the sequence. 
Thus, if CMAKE_FIND_FRAMEWORK is LAST it will find a dylib that appears anywhere in the sequence before finding a framework. So it will find /usr/local/lib/libvulkan.dylib if global SDK installation was chosen.

Reply:
If you're building with macOS and iOS, then you can use source VULKAN_SDK/iOS/setup_env.sh and source VULKAN_SDK/macOS/setup_env.sh  in the build process.
This is kinda a mess sadly and will get worse when we get watchOS or visionOS also setup this way.  However, this is what the recommended path is towards getting it to work.  I would prefer to use your solution, but we need to demonstrate the "official" solution path.  We'll cover this in the samples meeting today and see what the group thinks of this.

Reply:
> If you're building with macOS and iOS, then you can use source VULKAN_SDK/iOS/setup_env.sh and source VULKAN_SDK/macOS/setup_env.sh in the build process.

You can't if you are using the CMake GUI and Xcode IDE which I do. A single value of VULKAN_SDK is essential. It is easy enough for FindVulkan.cmake to do the right thing with knowledge of where in the SDK a single $VULKAN_SDK points.

Reply:
FindVulkan.cmake like other Find*.cmake solutions are meant to be project specific the way cmake architecture works.  So if the standard method doesn't work locally, it's totally fine to make changes for your work setup.

The reason we're pushing back here is because this is meant to be a standard solution.  If we make changes to FindVulkan.cmake then it gets those same changes to ALL other platforms.  *maybe* something that makes sense is to pickup a define from an option that is passed to cmake and if that option is set; have it update the environment path variable (or even check both the environment variable and the define?).  However, we can't make changes *just* for one platform without testing all platforms.  This requirement also includes testing on distributions that don't necessarily come from LunarG as the pathing wont always be the same; but the environment variables will be (that's why there's the environment variables).

To be blunt, this problem goes away when the config variation of find_package gets included with the SDK.  That's a larger task that will require much more than changes to the samples repo.  However, eventually FIndVulkan.cmake will no longer exist and vulkanconfig.cmake will be included in the SDK (and hopefully all SDKs).